homeassistant/components/viam/config_flow.py

```diff
@@ -0,0 +1,191 @@
+        vol.Required(CONF_ADDRESS): str,
+        vol.Required(CONF_SECRET): str,
+    }
+)
+STEP_AUTH_ORG_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_ID): str,
+        vol.Required(CONF_API_KEY): str,
+    }
+)
+
+
+async def validate_input(data: dict[str, Any]) -> tuple[str, ViamClient]:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    credential_type = data[CONF_CREDENTIAL_TYPE]
+    auth_entity = data.get(CONF_API_ID)
+    secret = data.get(CONF_API_KEY)
+    if credential_type == CRED_TYPE_LOCATION_SECRET:
+        auth_entity = data.get(CONF_ADDRESS)
+        secret = data.get(CONF_SECRET)
+
+    if not secret:
+        raise CannotConnect
+
+    creds = Credentials(type=credential_type, payload=secret)
+    opts = DialOptions(auth_entity=auth_entity, credentials=creds)
+    client = await ViamClient.create_from_dial_options(opts)
+
+    # If you cannot connect:
+    # throw CannotConnect
+    if client:
+        locations = await client.app_client.list_locations()
+        location = await client.app_client.get_location(next(iter(locations)).id)
+
+        # Return info that you want to store in the config entry.
+        return (location.name, client)
+
+    raise CannotConnect
+
+
+class ViamFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for viam."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize."""
+        self._title = ""
+        self._client: ViamClient
+        self._data: dict[str, Any] = {}
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            self._data.update(user_input)
+            return await self.async_step_auth()
+
+        return self.async_show_form(
+            step_id="user", data_schema=STEP_AUTH_USER_DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_auth(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
+                self._data.update(user_input)
+                (title, client) = await validate_input(self._data)
+                self._title = title
+                self._client = client
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return await self.async_step_robot()
+
+        schema = STEP_AUTH_ROBOT_DATA_SCHEMA
+        if self._data.get(CONF_CREDENTIAL_TYPE) == CRED_TYPE_API_KEY:
+            schema = STEP_AUTH_ORG_DATA_SCHEMA
```

Comment:
Instead of this, make two separate steps and jump to the correct step from the user step based on the selected credential type.

Reply:
Can I ask why? They would essentially be the same methods with just the set schema being different.

Reply:
It makes the code easier to follow. The two step handlers can fall back to some shared method if that helps deduplicate the code.

Reply:
Split the step and created a shared fallback method, which also helps keep the code coverage at 100% without needing redundant tests. Thanks for the feedback!